src/Symfony/Component/Config/Builder/ConfigBuilderGenerator.php

```diff
@@ -0,0 +1,360 @@
+
+            return new $className();
+        });
+
+        return $loader;
+    }
+
+    private function writeClasses(string $outputDir)
+    {
+        foreach ($this->classes as $class) {
+            $this->buildConstructor($class);
+            $this->buildToArray($class);
+            $dir = $outputDir.\DIRECTORY_SEPARATOR.$class->getFilePath();
+            @mkdir($dir, 0777, true);
+            file_put_contents($dir.\DIRECTORY_SEPARATOR.$class->getFilename(), $class->build());
+        }
+
+        $this->classes = [];
+    }
+
+    private function buildNode(NodeInterface $node, ClassBuilder $class, string $namespace)
+    {
+        if (!$node instanceof ArrayNode) {
+            throw new \LogicException('The node was expected to be an ArrayNode. This Configuration includes an edge case not supported yet.');
+        }
+
+        foreach ($node->getChildren() as $child) {
+            switch (true) {
+                case $child instanceof ScalarNode:
+                    $this->handleScalarNode($child, $class);
+                    break;
+                case $child instanceof PrototypedArrayNode:
+                    $this->handlePrototypedArrayNode($child, $class, $namespace);
+                    break;
+                case $child instanceof VariableNode:
+                    $this->handleVariableNode($child, $class);
+                    break;
+                case $child instanceof ArrayNode:
+                    $this->handleArrayNode($child, $class, $namespace);
+                    break;
+                default:
+                    throw new \RuntimeException(sprintf('Unknown node "%s".', \get_class($child)));
+            }
+        }
+    }
+
+    private function handleArrayNode(ArrayNode $node, ClassBuilder $class, string $namespace)
+    {
+        $childClass = new ClassBuilder($namespace, $node->getName());
+        $class->addRequire($childClass);
+        $this->classes[] = $childClass;
+
+        $property = $class->addProperty($node->getName(), $childClass->getName());
+        $body = '
+public function NAME(array $value = []): CLASS
+{
+    if (null === $this->PROPERTY) {
+        $this->PROPERTY = new CLASS($value);
+    } elseif ([] !== $value) {
+        throw new \Symfony\Component\Config\Definition\Exception\InvalidConfigurationException(sprintf(\'The node created by "NAME()" has already been initialized. You cannot pass values the second time you call NAME().\'));
+    }
+
+    return $this->PROPERTY;
+}';
+        $class->addMethod($node->getName(), $body, ['PROPERTY' => $property->getName(), 'CLASS' => $childClass->getFqcn()]);
+
+        $this->buildNode($node, $childClass, $namespace.'\\'.$childClass->getName());
+    }
+
+    private function handleVariableNode(VariableNode $node, ClassBuilder $class)
+    {
+        $comment = $this->getComment($node);
+        $property = $class->addProperty($node->getName());
+
+        $body = '
+/**
+COMMENT */
+public function NAME($value): self
```

Comment:
Im not 100%, but here is my thinking about default values. 
I've not added default values to properties, because that would make it impossible for the Config definition to know if the user set that value or left it as blank. 

To use default values in setters would make sense for the reasons you describe, I'll add it.